README.md
# @untemps/loan-schedule

Class to generate a schedule of monthly loan payments.

It has been built with a single purpose in mind so far, but features will be added in the next fews weeks.  
Don't hesitate to contact the team if you think about a specific usage we may address.

![npm](https://img.shields.io/npm/v/@untemps/loan-schedule?style=for-the-badge)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/untemps/loan-schedule/deploy?style=for-the-badge)
![Codecov](https://img.shields.io/codecov/c/github/untemps/loan-schedule?style=for-the-badge)

## Installation

```bash
yarn add @untemps/loan-schedule
```

## Usage

Import `LoanSchedule`:

```javascript
import { LoanSchedule } from '@untemps/loan-schedule'
```

Create an instance of `LoanSchedule` passing three objects:

```javascript
// Settings object that describes the properties of the loan (see below "Loan Settings")
const loanSettings = {
	amount: 79950, // Amount of the loan in any currency
	term: 25, // Loan term in years
	interestRate: 2, // Interest rate in percentage
	insuranceRate: 0.5, // Insurance rate in percentage
}

// List of amortizations monthly paid until the term of the loan (see below "Loan Amortizations")
const loanAmortizations = [
	{
		label: 'Rent', // Label of the amortization.
		value: 600, // Amount of the maortization.
		type: LoanScheduleAmortizationType.CREDIT, // Type of the amortization: "credit" or "debit"
	},
	{
		label: 'Payment',
		type: LoanScheduleAmortizationType.DEBIT,
		compute: LoanScheduleCompute.PAYMENT, // Computation method of the amortization (optional). That means the amount has to be automatically calculated by the class. "payment" and "insurance" are available so far.
	},
	{
		label: 'Property Tax',
		value: 30,
		rate: 5, // Yearly inflation rate of the amortization (optional).
		type: LoanScheduleAmortizationType.DEBIT,
	},
]

// Configuration of the output (see below "Configuration")
const config = {
    headers: {
        year: {
            label: 'Année',
        },
        balance: {
            label: 'Solde',
        },
    },
}

const schedule = new LoanSchedule(loanSettings, loanAmortizations, config);
```

### Constructor parameters

| Parameter             | Type                        | Description                                                                                        |
| --------------------- | --------------------------- | ---------------------------------------------------------------------------------------------------|
| loanSettings          | LoanScheduleSettings        | Settings object that describes the properties of the loan (see below **"Loan Settings"**)            |
| loanAmortizations     | LoanScheduleAmortization[]  | List of amortizations monthly paid until the term of the loan (see below **"Loan Amortizations"**)   |
| config                | LoanScheduleConfig          | Configuration of the output (see below **"Configuration"**)                                          |

### Loan Settings

Settings object that describes the properties of the loan

#### LoanScheduleSettings type

| Property              | Type                        | Description                                                                                        |
| --------------------- | --------------------------- | ---------------------------------------------------------------------------------------------------|
| amount                | number                      | Amount of the loan in any currencyAmount of the loan in any currency                               |
| term                  | number                      | Loan term in years                                                                                 |
| interestRate          | number                      | Interest rate in percentage                                                                        |
| insuranceRate         | number                      | Insurance rate in percentage                                                                       |

### Loan Amortizations

List of amortizations monthly paid until the term of the loan

#### LoanScheduleAmortization type

| Property              | Type                          | Description                                                                                                                                                                |
| --------------------- | ----------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| label                 | string                        | Label of the amortization                                                                                                                                                  |
| value                 | number                        | Monthly amount of the amortization                                                                                                                                         |
| rate                  | number                        | Yearly inflation rate of the amortization (optional).                                                                                                                      |
| type                  | LoanScheduleAmortizationType  | Type of the amortization: "credit" or "debit"                                                                                                                              |
| compute               | LoanScheduleCompute           | Computation method of the amortization (optional). That means the amount has to be automatically calculated by the class. "payment" and "insurance" are available so far.  |

### Configuration

The configuration object allows to change the output to reflect your specific needs.

#### LoanScheduleConfig type

| Property              | Type                          | Description                                             |
| --------------------- | ----------------------------- | --------------------------------------------------------|
| headers               |                               |                                                         |
| --- payment           |                               |                                                         |
| ------- label         | string                        | Label of the payment amortization                       |
| --- insurance         |                               |                                                         |
| ------- label         | string                        | Label of the insurance amortization                     |

## Development

A demo can be served for development purpose on `http://localhost:10001/` running:

```
yarn dev
```


CHANGELOG.md
## [2.1.2](https://github.com/untemps/loan-schedule/compare/v2.1.1...v2.1.2) (2021-04-21)

## [2.1.1](https://github.com/untemps/loan-schedule/compare/v2.1.0...v2.1.1) (2021-04-21)

# [2.1.0](https://github.com/untemps/loan-schedule/compare/v2.0.0...v2.1.0) (2021-04-15)


### Features

* Export LoanScheduleYears type ([299eca7](https://github.com/untemps/loan-schedule/commit/299eca71b4377de3622c6a9e38b239b02cae5cca))

# [2.0.0](https://github.com/untemps/loan-schedule/compare/v1.0.1...v2.0.0) (2021-04-14)


### Bug Fixes

* Delete build watch mode ([acfdd39](https://github.com/untemps/loan-schedule/commit/acfdd3949bf6910ce3e676c7df8e7ba380b32206))
* Fix unit tests ([e095435](https://github.com/untemps/loan-schedule/commit/e0954358ee4ee2fc372f084acae39b8445fb4d21))


### Code Refactoring

* Refactor lib architecture ([8122250](https://github.com/untemps/loan-schedule/commit/8122250045fd7261be7f583f75a9e045a675b32b))


### BREAKING CHANGES

* LoanSchedule signature and output are very different from previous version

## [1.0.1](https://github.com/untemps/loan-schedule/compare/v1.0.0...v1.0.1) (2021-04-12)

# 1.0.0 (2021-04-05)


### Features

* Initialize project ([e8daf82](https://github.com/untemps/loan-schedule/commit/e8daf82284d28ad82f65a11c24e57311c7a0761a))


dev/src/index.js
import { LoanSchedule, LoanScheduleCompute, LoanScheduleAmortizationType } from '../../dist'

import draw from './draw'

const schedule = new LoanSchedule(
	{
		amount: 79950,
		term: 25,
		interestRate: 2,
		insuranceRate: 0.5,
	},
	[
		{
			label: 'Loyer',
			value: 600,
			type: LoanScheduleAmortizationType.CREDIT,
		},
		{
			label: 'Mensualité',
			type: LoanScheduleAmortizationType.DEBIT,
			compute: LoanScheduleCompute.PAYMENT,
		},
		{
			label: 'Assurance',
			type: LoanScheduleAmortizationType.DEBIT,
			compute: LoanScheduleCompute.INSURANCE,
		},
		{
			label: 'Taxe foncière',
			value: 30,
			rate: 5,
			type: LoanScheduleAmortizationType.DEBIT,
		},
		{
			label: 'Charges de copro',
			value: 40,
			rate: 5,
			type: LoanScheduleAmortizationType.DEBIT,
		},
	],
	{
		headers: {
			year: {
				label: 'Année',
			},
			balance: {
				label: 'Solde',
			},
		},
	}
)
console.log(schedule.years)
draw(schedule.years)
